src/invariant/OrderBookIsNotCrossed.cpp

```diff
@@ -0,0 +1,225 @@
+                    : std::make_pair(offer.buying, offer.selling);
+
+            assets.insert(assetPair);
+        }
+    }
+
+    return assets;
+}
+
+static double
+price(OfferEntry const& offer)
+{
+    return double(offer.price.n) / double(offer.price.d);
+}
+
+static std::string
+checkCrossed(Asset const& a, Asset const& b, OrderBook const& orderBook)
+{
+    // if either side of order book for asset pair empty or does not yet exist,
+    // order book cannot be crossed
+    auto const iterAB = orderBook.find({a, b});
+    auto const iterBA = orderBook.find({b, a});
+    if (iterAB == orderBook.end() || iterBA == orderBook.end())
+    {
+        return {};
+    }
+
+    auto const& asks = iterAB->second;
+    auto const& bids = iterBA->second;
+    if (asks.empty() || bids.empty())
+    {
+        return {};
+    }
+
+    // check if crossed
+    auto lowestAsk = price(*asks.cbegin());
+    auto highestBid = 1 / price(*bids.cbegin());
+
+    // Allow highestBid to equal (not exceed) lowestAsk if at least one of the
+    // offers is passive.
+    if (lowestAsk == highestBid)
+    {
+        auto existsEqualPriceNonPassive =
+            [](auto const targetPrice, Orders const& offers, auto& found) {
+                for (auto const& offer : offers)
+                {
+                    if (price(offer) != targetPrice)
+                    {
+                        return false;
+                    }
+                    if ((offer.flags & PASSIVE_FLAG) == 0)
+                    {
+                        found = offer;
+                        return true;
+                    }
+                }
+                return false;
+            };
+
+        OfferEntry ask, bid;
+        auto nonPassiveAskAtSamePrice =
+            existsEqualPriceNonPassive(lowestAsk, asks, ask);
+        auto nonPassiveBidAtSamePrice =
+            existsEqualPriceNonPassive(1 / lowestAsk, bids, bid);
+
+        if (nonPassiveAskAtSamePrice && nonPassiveBidAtSamePrice)
+        {
+            return fmt::format(
+                FMT_STRING("Order book is in a crossed state for {} - {} "
+                           "asset pair.\nEqual price {} for non-passive offers "
+                           "{} and {}."),
+                xdr::xdr_to_string(a), xdr::xdr_to_string(b), lowestAsk,
+                xdr::xdr_to_string(ask), xdr::xdr_to_string(bid));
+        }
+
+        return {};
+    }
+
+    if (lowestAsk < highestBid)
+    {
+        return fmt::format(
+            FMT_STRING(
+                "Order book is in a crossed state for {} - {} "
+                "asset pair.\nTop of the book is:\n\tAsk price: {}\n\tBid "
+                "price: {}\n\nWhere {} >= {}!"),
+            xdr::xdr_to_string(a), xdr::xdr_to_string(b), lowestAsk, highestBid,
+            highestBid, lowestAsk);
+    }
+    return {};
+}
+}
+
+void
+OrderBookIsNotCrossed::deleteFromOrderBook(OfferEntry const& oe)
```

Comment:
Nitpick: are these functions really adding any value? They're both one liners that are easy to understand, and they have a single caller.

Reply:
Agreed, I'll inline those.